### PR TITLE
separate infrastructure environments

### LINKED ISSUE #830: have a way to test Terraform changes more safely
Currently, our different environments are set up as App Service slots. This means infrastructure (a.k.a. Terraform) changes outside of slots apply to _all of them_, meaning there isn't an easy way to test those changes beyond a `plan`.

I propose we use a dedicated Resource Group (maybe also a dedicated subscription?) to create a full `staging` environment, perhaps with some features like the Monitors disabled. We might want to go all the way and make one for `dev`, and get rid of our use of slots. Open to feedback here.

cc @cal-itp/cdt-devsecops 

## Acceptance Criteria

- [ ] Changes can be made to Terraform tested without concern that production will be impacted

## Additional context

This would have helped prevent [the 7/21 downtime](https://docs.google.com/document/d/14XFpJonXmrNKDt98g8vML3RimbZha0UJiIeUVclXYGs/edit#).

### PR DESCRIPTION
**Pull request by afeld**
Closes #830. Running list of TODOs:

- [x] Update docs around environments, removing references to "slots"
- [x] Update deployment pipeline
  - [x] Ensure it's connecting to the right subscription/backend
- [ ] DevSecOps:

  - [x] [Recover or purge Key Vault](https://learn.microsoft.com/en-us/azure/key-vault/general/key-vault-recovery?tabs=azure-portal#list-recover-or-purge-a-soft-deleted-key-vault) `KV-CDT-PUB-CALITP-D-001`
    - Our permissions are limited on Key Vaults so have been trying to work around that
  - [x] Create VNet and Subnet(s) for dev and test
    - [x] Provide VNet and Subnet names, and the Resource Group they're in
    - [ ] Give read permissions:

      > Error: retrieving Subnet: (Name "SNET-CDT-PUB-CALITP-D-001" / Virtual Network Name "VNET-CDT-PUB-SHRD-W-D-001" / Resource Group "RG-CDT-PUB-SHRD-W-D-001"): network.SubnetsClient#Get: Failure responding to request: StatusCode=403 -- Original Error: autorest/azure: Service returned an error. Status=403 Code="AuthorizationFailed" Message="The client 'live.com#aidan@compiler.la' with object id \[REDACTED\] does not have authorization to perform action 'Microsoft.Network/virtualNetworks/subnets/read' over scope '/subscriptions/\[REDACTED\]/resourceGroups/RG-CDT-PUB-SHRD-W-D-001/providers/Microsoft.Network/virtualNetworks/VNET-CDT-PUB-SHRD-W-D-001/subnets/SNET-CDT-PUB-CALITP-D-001' or the scope is invalid.

  - [x] Create `Test` Resource Group
  - [x] Create Azure DevOps Service Connection to `Development` Subscription
  - [x] Configure Front Door
  - [ ] Clean up unused resources in `Development` Resource Group:
    - [ ] Storage account `sacalitpd001`
    - [x] App Service Plan `ASP-RGCDTPUBVIPCALITPD001-9628`
  - [x] Provide Active Directory Group ID

- [x] [Rename resources](https://developer.hashicorp.com/terraform/language/modules/develop/refactoring) to:
  - Prevent them from being destroyed and recreated
  - Make the `plan` output shorter and thus easier to review
- [x] Import `test` resources
- [ ] Update DNS
  - [ ] Set hostname verification `TXT` record
    - [x] dev
    - [ ] test
  - [ ] [test-benefits.calitp.org](http://test-benefits.calitp.org/) CNAME [FD-CDT-SHRD-D-001.azurefd.net](http://fd-cdt-shrd-d-001.azurefd.net/)
  - [ ] [dev-benefits.calitp.org](http://dev-benefits.calitp.org/) CNAME [FD-CDT-SHRD-D-001.azurefd.net](http://fd-cdt-shrd-d-001.azurefd.net/)
- [x] Set up webhooks from GitHub
- [x] Set configuration variables and files for new environments
- [ ] https://github.com/cal-itp/benefits/issues/1142
- [ ] Update IP address ranges in Eligibility Server
  - [x] dev
  - [ ] test - submitted https://github.com/cal-itp/eligibility-server/pull/233 to help with that

### CODE REVIEW

File: terraform/main.tf [hunk 1]
```diff
@@ -4,7 +4,7 @@ terraform {
   required_providers {
     azurerm = {
       source  = "hashicorp/azurerm"
-      version = "~> 3.25.0"
+      version = ">= 3.0.0, < 4.0.0"
```
Comment (afeld) [inline terraform/main.tf:7]: The version gets locked anyway, so no need to have a tighter constraint.

File: terraform/main.tf [hunk 2]
```diff
@@ -22,7 +22,3 @@ provider "azurerm" {
 }
 
 data "azurerm_client_config" "current" {}
-
-data "azurerm_resource_group" "prod" {
-  name = "RG-CDT-PUB-VIP-CALITP-P-001"
-}
```
Comment (afeld) [inline terraform/main.tf:28]: Moved to `environment.tf`.

File: terraform/network.tf [hunk 3]
```diff
@@ -1,10 +1,8 @@
...
+  network_resource_group_name = local.is_prod ? "RG-CDT-PUB-SHRD-W-P-001" : "RG-CDT-PUB-D-001"
+  vnet_name                   = local.is_prod ? "VNET-CDT-PUB-SHRD-W-P-001" : "VNET-CDT-PUB-D-001"
+  subnet_name                 = local.is_prod ? "SNET-CDT-PUB-CALITP-P-001" : "SN-CDT-PUB-CALITP-${local.env_letter}-001"
 
-data "azurerm_subnet" "main" {
-  name                 = "SNET-CDT-PUB-CALITP-P-001"
-  virtual_network_name = "VNET-CDT-PUB-SHRD-W-P-001"
-  resource_group_name  = local.network_resource_group_name
+  subnet_id = "/subscriptions/${data.azurerm_client_config.current.subscription_id}/resourceGroups/${local.network_resource_group_name}/providers/Microsoft.Network/virtualNetworks/${local.vnet_name}/subnets/${local.subnet_name}"
 }
```
Comment (afeld) [inline terraform/network.tf:8]: Removes the need to have read access to the subnet.

**Comment by afeld** [comment]
The Terraform here is all ready to go, thus changes can be deployed manually (with `terraform apply`). The Pipeline changes are on hold pending https://github.com/cal-itp/benefits/issues/830#issuecomment-1347213635.

**Comment by afeld** [comment]
We are now seeing the pipeline working appropriately; see the [run against `dev`](https://calenterprise.visualstudio.com/CDT.OET.CAL-ITP/_build/results?buildId=38573&view=results) and the [run against `prod`](https://calenterprise.visualstudio.com/CDT.OET.CAL-ITP/_build/results?buildId=38574&view=results), each of which is using the appropriate service connection, workspace, etc.

File: terraform/network.tf [hunk 4]
```diff
@@ -1,10 +1,9 @@
 locals {
-  # VNet uses a shared resource group different from the main App Service resource group
-  network_resource_group_name = "RG-CDT-PUB-SHRD-W-P-001"
-}
+  # VNet uses a shared Resource Group, different from App Service Resource Group we use for Benefits stuff
+  network_resource_group_name = local.is_prod ? "RG-CDT-PUB-SHRD-W-P-001" : "RG-CDT-PUB-D-001"
+  vnet_name                   = local.is_prod ? "VNET-CDT-PUB-SHRD-W-P-001" : "VNET-CDT-PUB-D-001"
+  subnet_name                 = local.is_prod ? "SNET-CDT-PUB-CALITP-P-001" : "SN-CDT-PUB-CALITP-${local.env_letter}-001"
```
Comment (thekaveman) [inline terraform/network.tf:5]: Is the right side of this ternary correct? `"SN-CDT-PUB-CALITP-${local.env_letter}-001"`

Should it be `"SNET-CDT-PUB-CALITP-${local.env_letter}-001"`

File: terraform/pipeline/workspace.py [hunk 5]
```diff
@@ -0,0 +1,46 @@
+"""Used to set the environment-related variables at runtime (rather than build
+time) so that all the necessary pipeline variables are available."""
+
+import os
+import sys
+
+REASON = os.environ["REASON"]
+# the name of the variable that Azure Pipelines uses for the source branch
+# depends on the type of run, so need to check both
+SOURCE = os.environ.get("OTHER_SOURCE") or os.environ["INDIVIDUAL_SOURCE"]
```
Comment (thekaveman) [inline terraform/pipeline/workspace.py:10]: Very minor, take it or leave it

```suggestion
SOURCE = os.environ.get("OTHER_SOURCE", os.environ["INDIVIDUAL_SOURCE"])
```

File: terraform/pipeline/workspace.py [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+TARGET = os.environ["TARGET"]
+
+# the branches that correspond to environments
+ENV_BRANCHES = ["dev", "test", "prod"]
+
+if REASON == "PullRequest" and TARGET in ENV_BRANCHES:
+    # it's a pull request against one of the environment branches, so use the
+    # target branch
+    environment = TARGET
+elif REASON == "IndividualCI" and SOURCE in ENV_BRANCHES:
```
Comment (thekaveman) [inline terraform/pipeline/workspace.py:20]: Do we want the same `"Manual"` reason here, like from cal-itp/eligibility-server#233

File: terraform/network.tf (hunk 4 above)
Comment (afeld) [reply terraform/network.tf:5]: It matches what @cal-itp/cdt-devsecops named it 🤷

File: terraform/pipeline/workspace.py (hunk 5 above)
Comment (afeld) [reply terraform/pipeline/workspace.py:10]: Thanks, but I think it's clearer as written.

File: terraform/pipeline/workspace.py (hunk 6 above)
Comment (afeld) [reply terraform/pipeline/workspace.py:20]: Ah yes, thanks, I forgot this script was only introduced by this pull request.

File: terraform/network.tf (hunk 4 above)
Comment (afeld) [reply terraform/network.tf:5]: As part of https://github.com/cal-itp/benefits/issues/1077, maybe we can raise where we are seeing inconsistencies.

**Review by angela-tran — APPROVED** [review]

**Comment by afeld** [comment]
I had already run a `terraform apply` locally with these changes, so [`No changes` from the Pipeline's `apply` on `dev`](https://calenterprise.visualstudio.com/CDT.OET.CAL-ITP/_build/results?buildId=38588&view=logs&j=5af648c4-4294-55a7-6072-dd70820d3dba&t=39708da4-e251-5130-285a-f2fe36c9a5aa&l=42). Expect to see changes as we merge that up to `test`, and then the merge to `prod` will be the big one that deletes the slots, etc.

**Pull request merged**